[IRGen] Lowered open_pack_element.

nate-chandler opened the pull request:
Initial lowering support for the open_pack_element instruction.  While lowering,

```
open_pack_element 
  %index 
  of <..., U_i..., ...> 
  at <..., Pack_i, ....>, 
  shape $U_k, 
  uuid "01234567-89AB-CDEF-0123-000000000000"
```

binds `(@pack_element(...) U_i)` to the type `T(i, %index)` at `%index` in `Pack_i` for all `i` of the appropriate shape.

Materialize the metadata address for `T(i, %index)` depending on what's already available.  In general, compare the %index to the upper bound of each element's range: if it's in range, emit the metadata at `%innerIndex := %index - %innerLowerBound`; if it's out of range, jump to the next comparison.  From each metadata emission block, jump to a merge point where the result metadata is defined to be the phi of the metadata materialized in each predecessor.

Still to do: handle conformances.

nate-chandler committed:
[NFC] Extracted getForwardedPackArchetype.

nate-chandler committed:
[NFC] Extracted tryGetLocalPackTypeMetadata.

nate-chandler committed:
[NFC] GenericRequirement vends an llvm::Type *.

The mapping from requirement kind to llvm::Type is repeated several
places.  Pull it into a helper.

nate-chandler committed:
[NFC] Extracted emitPackExpansionElementMetadata.

The function emits the metadata for the element of a pack expansion at a
specified index.  It was previously defined inline in the loop--defined
in emitPackExpansionType--which populates a new metadata array with
elements derived from a preexisting array.  It will be used to emit a
single element in the case that only one is needed.

nate-chandler committed:
[IRGen] Lowered open_pack_element.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

rjmccall commented on lib/IRGen/GenPack.cpp:
```diff
@@ -164,30 +164,9 @@ static Address emitFixedSizeMetadataPackRef(IRGenFunction &IGF,
-  IGF.Builder.emitBlock(loop);
-
+static llvm::Value *emitPackExpansionElementMetadata(
+    IRGenFunction &IGF, CanPackExpansionType expansionTy, llvm::Value *phi,
```
`phi` is an unfortunate name for a parameter, since you don't care whether it's a phi.  `index`, maybe?

rjmccall commented on lib/IRGen/GenProto.cpp:
```diff
@@ -3618,6 +3618,24 @@ void irgen::bindFromGenericRequirementsBuffer(IRGenFunction &IGF,
+  case GenericRequirement::Kind::Metadata: {
+    CanGenericTypeParamType ty;
+    if ((ty = dyn_cast<GenericTypeParamType>(requirement.getTypeParameter())) &&
+        ty->isParameterPack())
```
We might see this kind of requirement for associated types of packs, too, right?

You should have enough information to set the requirement kind as `MetadataPack` / `WitnessTablePack` in the first place, and that'll probably make clients of this a lot easier.

nate-chandler replied:
Added these new kinds in https://github.com/apple/swift/pull/63363/commits/6a44de214ddaafb216c8fd85acb9bb79bd6ee296 .  Does the static member function `GenericRequirement::isPack` added there correctly identify which types are packs?

rjmccall commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -6863,8 +6864,17 @@ void IRGenSILFunction::visitScalarPackIndexInst(ScalarPackIndexInst *i) {
+                                                    MetadataState::Complete);
+    this->bindArchetype(archetype, metadata, MetadataState::Complete, {});
+    ++packIndex;
+  });
```
I'm not sure I see why this is being done in two loops.

nate-chandler replied:
Fixed.

nate-chandler committed:
[NFC] Improved parameter name.

The parameter name to emitPackExpansionElementMetadata matched the
preexisting local variable name.  Change it to something meaningful to
callers.  Use a separate commit to make the original extraction more
obviously just an extraction.

nate-chandler committed:
[GenericRequirement] Added pack variants.

Added pack flavors of requirement kinds for metadata and witness tables.

Fixes the function signatures for variadic generic functions which
previously used %swift.type* for variadic generic parameters--those are
lists of metadata and should actually be %swift.type**.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test macOS platform
